### PR TITLE
Subtype Parser

### PR DESCRIPTION
**Pull request by jovianyu**
Initial script for adding streamlining subtype information as well as
core events, with test input.

### CODE REVIEW

File: subtype-parser/subtype.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    with open( mutationMatrix ) as f:
+        MM = [ l.rstrip().split( '\t' ) for l in f if not l.startswith( '#' ) ]
+
+    # Generating output for new mutation matrix with added subtype information
+    if tempDirectory[-1] != '/':
+        tempDirectory += '/'
+    with open( tempDirectory + 'temp.snv', "w" ) as out:
+        output = []
+        for p in MM:
+            if subtypeDict[p[0]] != None:
```
Comment (mdml) [inline subtype-parser/subtype.py]: This is not quite right. If patient _P_ is of subtype _T_, we want to add events (aberrations) for all subtypes **except** _T_. That way genes with alterations that are mutually exclusive with the subtype _T_ alteration are actually enriched for mutations in _T_.

Comment (bournewu) [inline subtype-parser/subtype.py]: Line 59 should change to only adding subtypes that are not in the patient P. For example, 

``` python
p.extend( [ s for s in subtypes.difference( set([subtypeDict[p[0]]]) ) ])
```

**Comment by jovianyu** [comment]
Subtype parser now fully merged into the CoMEt codebase and verified on test data.

File: comet/mutation_data.py [hunk 2]
```diff
@@ -20,31 +20,49 @@ def load_mutation_data(filename, patientFile=None, geneFile=None, minFreq=0):
...
     # Parse the mutation matrix
-    from collections import defaultdict
     geneToCases, patientToGenes = defaultdict(set), defaultdict(set)
     with open(filename) as f:
         arrs = [ l.rstrip().split("\t") for l in f if not l.startswith("#") ]
         for arr in arrs:
             patient, mutations = arr[0], set(arr[1:])
 
+            if subtypeDict[patient]:
+                mutations = mutations.union( subtypes.difference( set( [subtypeDict[patient]] ) ) )
```
Comment (mdml) [inline comet/mutation_data.py]: I would write something like the following instead:

```
mutations |= set( s for s in subtypes if subtypeDict[patient] != s )
```

File: comet/mutation_data.py [hunk 3]
```diff
@@ -20,31 +20,49 @@ def load_mutation_data(filename, patientFile=None, geneFile=None, minFreq=0):
...
+
             if not patients or patient in patients:
                 if genes: mutations &= genes
-                #else: genes |= mutations                
+                #else: genes |= mutations
```
Comment (mdml) [inline comet/mutation_data.py]: Why is this commented out? Should it be deleted?

File: run_comet.py [hunk 4]
```diff
@@ -246,14 +250,15 @@ def run( args ):
...
-    mutations = C.load_mutation_data(mutationMatrix, patientFile, geneFile, minFreq)
-    m, n, genes, patients, geneToCases, patientToGenes = mutations
+    mutations = C.load_mutation_data(mutationMatrix, patientFile, geneFile, minFreq, subtypeFile)
+    m, n, genes, patients, geneToCases, patientToGenes, subtypes = mutations
+    mutations = ( m, n, genes, patients, geneToCases, patientToGenes )
 
-    if args.subtype:
-        with open(args.subtype) as f:
-            subSet = [ l.rstrip() for l in f ]
+    if args.core_events:
```
Comment (mdml) [inline run_comet.py:258]: This looks good, but can you add some comments to explain the logic here?

**Comment by mdml** [comment]
This looks good to me once you address the few comments I made.

@bournewu: can you review and test Jovian's pull request?

**Comment by bournewu** [comment]
Just reviewed and tested. Now looks good to me. @jovianyu could you merge it?

Just merged to the master. : )

**Pull request merged**